src/main/java/io/github/netmikey/mvncloner/mvncloner/Publisher.java

```diff
@@ -39,63 +39,86 @@ public class Publisher {
-        throws IOException, InterruptedException {
+        throws IOException, ExecutionException, InterruptedException {
 
         LOG.debug("Switching to mirror directory: " + mirrorPath.toAbsolutePath());
 
-        List<Path> recursePaths = new ArrayList<>();
-
+        List<Path> recursePaths = new LinkedList<>();
+        Collection<Future<?>> futures = new HashSet<>();
         try (DirectoryStream<Path> stream = Files.newDirectoryStream(mirrorPath)) {
             for (Path path : stream) {
                 if (Files.isDirectory(path)) {
                     recursePaths.add(path);
                 } else {
-                    handleFile(httpClient, repositoryUrl, path);
+                    futures.add(runner.submit(() -> handleFile(httpClient, repositoryUrl, path)));
                 }
             }
         }
 
+        for (var future : futures) {
+            future.get();
+        }
+
         // Tail recursion
         for (Path recursePath : recursePaths) {
             String subpath = mirrorPath.relativize(recursePath).toString();
             publishDirectory(httpClient, appendUrlPathSegment(repositoryUrl, subpath), recursePath);
         }
     }
 
-    private void handleFile(HttpClient httpClient, String repositoryUrl, Path path)
-        throws IOException, InterruptedException {
+    private void handleFile(HttpClient httpClient, String repositoryUrl, Path path) {
 
         String filename = path.getFileName().toString();
         String targetUrl = repositoryUrl + filename;
-        LOG.info("Uploading " + targetUrl);
-
-        Utils.sleep(1000);
-        HttpRequest request = Utils.setCredentials(HttpRequest.newBuilder(), username, password)
-            .uri(URI.create(targetUrl))
-            .PUT(BodyPublishers.ofInputStream(() -> {
-                try {
-                    return Files.newInputStream(path, StandardOpenOption.READ);
-                } catch (IOException e) {
-                    throw new RuntimeException(e);
-                }
-            }))
-            .build();
-        HttpResponse<String> response = httpClient.send(request, BodyHandlers.ofString());
-        if (response.statusCode() < 200 || response.statusCode() > 299) {
-            LOG.error("Error uploading " + targetUrl + " : Response code was " + response.statusCode());
-            LOG.debug("   Response headers: " + response.headers());
-            LOG.debug("   Response body: " + response.body());
+        try {
+            var baseReq = Utils.setCredentials(HttpRequest.newBuilder(), username, password)
+                .uri(URI.create(targetUrl));
+
+            var getResponse = httpClient.send(baseReq.method("HEAD", BodyPublishers.noBody()).build(), BodyHandlers.discarding());
+            if (getResponse.statusCode() != 404) {
+                LOG.info("Artifact {} already exists", targetUrl);
+                LOG.debug("   Response headers: " + getResponse.headers());
+                return;
+            }
```

Comment:
Why did you add an additional http HEAD request? Not sure if this is needed and  backwards compatible? Also, assuming anything but http 404 means the file already exists is probably not correct.

Reply:
This is to be more idempotent, in large process some fails occours (here about 3 days of sync) and re-upload all data only to get a fail isn't fast nor efficient. 
with this, if the file already exists, the upload no longer is executed

Reply:
Ok I see.

Would you mind making it configurable though? I could imagine that there's a use-case out there where people want to explicitly overwrite all or parts of an already existing repository. In such case, people could cut the amount of http requests in half by omitting the http HEAD request and PUT'ting immediately.

Something like: `--target.skip-existing=true`.